<a href="https://colab.research.google.com/github/sebekpro/python-krakowiak/blob/main/250-data-science-bootcamp-w-jezyku-python/061_070_exercises.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Numpy

### Spis treści:
* [Import biblioteki](#0)
* [Ćwiczenie 61](#1)
* [Ćwiczenie 62](#2)
* [Ćwiczenie 63](#3)
* [Ćwiczenie 64](#4)
* [Ćwiczenie 65](#5)
* [Ćwiczenie 66](#6)
* [Ćwiczenie 67](#7)
* [Ćwiczenie 68](#8)
* [Ćwiczenie 69](#9)
* [Ćwiczenie 70](#10)

### <a name='0'></a>Import biblioteki

In [2]:
import numpy as np

np.set_printoptions(suppress=True, precision=4)
np.random.seed(42)
np.__version__

'1.26.4'

### <a name='1'></a> Ćwiczenie 61
Podane są poniższe dwie tablice $A$ i $B$:

```
A = np.array([[3, 4, 9, 2],
              [5, 3, 2, 5]])
B = np.array([[4, 3, 2, 5],
              [6, 3, 1, 6]])
```

Wyznacz tablicę składającą się ze średnich arytmetycznych tablic $A$ i $B$ odpowiadających im elementów.

__Oczekiwany wynik:__


```
array([[3.5, 3.5, 5.5, 3.5],
       [5.5, 3. , 1.5, 5.5]])
```



In [3]:
# tutaj wpisz rozwiązanie
A = np.array([[3, 4, 9, 2],
              [5, 3, 2, 5]])
B = np.array([[4, 3, 2, 5],
              [6, 3, 1, 6]])
print ((A + B) / 2)

[[3.5 3.5 5.5 3.5]
 [5.5 3.  1.5 5.5]]


### <a name='2'></a> Ćwiczenie 62
Podane są poniższe dwie tablice $A$ i $B$:

```
A = np.array([[3, 4, 9, 2],
              [5, 3, 2, 5]])
B = np.array([[4, 3, 2, 5],
              [6, 3, 1, 6]])
```
Pomnóż tablice przez siebie - element po elemencie (element-wise).


In [4]:
# tutaj wpisz rozwiązanie
A = np.array([[3, 4, 9, 2],
              [5, 3, 2, 5]])
B = np.array([[4, 3, 2, 5],
              [6, 3, 1, 6]])

C = A * B
C

array([[12, 12, 18, 10],
       [30,  9,  2, 30]])

### <a name='3'></a> Ćwiczenie 63
Podana jest poniższa tablica $A$:

```
A = np.array([[3, 4, 9, 2],
              [5, 3, 2, 5]])
```
Wyznacz pierwiastek z każdego elementu tablicy $A$.

__Wskazówka:__ Użyj funkcji _np.sqrt()_.



In [5]:
# tutaj wpisz rozwiązanie
A = np.array([[3, 4, 9, 2],
              [5, 3, 2, 5]])
B = np.sqrt(A)
B

array([[1.7321, 2.    , 3.    , 1.4142],
       [2.2361, 1.7321, 1.4142, 2.2361]])

### <a name='4'></a> Ćwiczenie 64
Podane są poniższe tablice $A$ oraz $B$:


```
A = np.linspace(0, np.pi / 2, 20)
B = np.full(shape=(20,), fill_value=1, dtype='float')
```
Sprawdź działanie jedynki trygonometrycznej na tablicy $A$ i następnie porównaj wynik z tablicą $B$. Użyj funkcji _np.allclose()_.

__Jedynka trygonometryczna:__ Dla dowolnej wartości kąta $\alpha \in R$ zachodzi:
$$sin^2\alpha + cos^2\alpha = 1$$



In [6]:
# tutaj wpisz rozwiązanie
A = np.linspace(0, np.pi / 2, 20)
B = np.full(shape=(20,), fill_value=1, dtype='float')

# Obliczamy sin^2(A) + cos^2(A) dla każdej wartości kąta w tablicy A
result = np.sin(A)**2 + np.cos(A)**2

# Porównujemy wynik z tablicą B, sprawdzając, czy elementy są "bliskie"
equal = np.allclose(result, B)

print(f"Tablica A: \n {A}\n")
print(f"Tablica B: \n {B}\n")
print("Wynik sin^2 + cos^2 dla A:")
print(result)
print("\nCzy wynik jest równy B (jedynce)?:", equal)

Tablica A: 
 [0.     0.0827 0.1653 0.248  0.3307 0.4134 0.496  0.5787 0.6614 0.7441
 0.8267 0.9094 0.9921 1.0748 1.1574 1.2401 1.3228 1.4054 1.4881 1.5708]

Tablica B: 
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]

Wynik sin^2 + cos^2 dla A:
[1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]

Czy wynik jest równy B (jedynce)?: True


### Przypomnienie
### Mnożenie macierzy:
$${A} ={\begin{bmatrix}a_{11}&a_{12}&\cdots &a_{1n}\\a_{21}&a_{22}&\cdots &a_{2n}\\\vdots &\vdots &\ddots &\vdots \\a_{m1}&a_{m2}&\cdots &a_{mn}\\\end{bmatrix}}$$
A - macierz wymairu $m\ x\ n$

$$ {B} ={\begin{bmatrix}b_{11}&b_{12}&\cdots &b_{1p}\\b_{21}&b_{22}&\cdots &b_{2p}\\\vdots &\vdots &\ddots &\vdots \\b_{n1}&b_{n2}&\cdots &b_{np}\\\end{bmatrix}}$$
B - macierz wymairu $n\ x\ p$

$$C = AB$$
C - macierz wymiaru $m\ x\ p$  
Aby móc pomnożyć dwie macierze liczba kolumn w lewej macierzy musi zgadzać się z liczbą wierszy prawej macierzy.

Mnożenie macierzy nie jest przemienne! Tzn. $$AB \neq BA$$

### Przykład 1
$ X = \begin{bmatrix}5&3\\3&9\end{bmatrix}$, $ Y= \begin{bmatrix}1\\-1\end{bmatrix} $  
$X$ macierz 2x2  
$Y$ macierz 2x1
$$Z = X \cdot Y = \begin{bmatrix}5&3\\3&9\end{bmatrix} \cdot \begin{bmatrix}1\\-1\end{bmatrix} = \begin{bmatrix}2\\-6\end{bmatrix}$$
$Z$ macierz 2x1

### Przykład 2
$X = \begin{bmatrix}2&-1&3\\3&1&0\end{bmatrix}$, $Y = \begin{bmatrix}2&1&-1\\0&-1&2\\3&2&0\end{bmatrix}$

$X$ macierz 2x3  
$Y$ macierz 3x3
$$Z = X \cdot Y =  \begin{bmatrix}2&-1&3\\3&1&0\end{bmatrix} \cdot \begin{bmatrix}2&1&-1\\0&-1&2\\3&2&0\end{bmatrix} = \begin{bmatrix}13&9&-4\\6&2&-1\end{bmatrix}$$
$Z$ macierz 2x3




### <a name='5'></a> Ćwiczenie 65

Podane są dwie dwuwymiarowe tablice numpy $A$ oraz $B$.  
$A$ - macierz wymiaru $(3, 2)$  
$B$ - macierz wymiaru $(2, 3)$

```
A = np.array([[2, 3],
              [-4, 2],
              [5, 0]])
B = np.array([[4, 3, 2],
              [-1, 0, 2]])
```
Dokonaj mnożenia macierzowego tablic $A$ i $B$.

__Wskazówka:__ Użyj funkcji _np.dot()_.


In [8]:
# tutaj wpisz rozwiązanie
A = np.array([[2, 3],
              [-4, 2],
              [5, 0]])
B = np.array([[4, 3, 2],
              [-1, 0, 2]])

C = np.dot(A, B)
C

array([[  5,   6,  10],
       [-18, -12,  -4],
       [ 20,  15,  10]])

### Przypomnienie

### Wyznacznik macierzy:

$${\displaystyle A={\begin{bmatrix}a_{11}&a_{12}&\dots &a_{1n}\\a_{21}&a_{22}&\dots &a_{2n}\\\vdots &\vdots &\ddots &\vdots \\a_{n1}&a_{n2}&\dots &a_{nn}\end{bmatrix}}}$$
$A$ - macierz wymiaru nxn

#### Oznaczenia:

${\displaystyle |A|=\left|{\begin{array}{c}a_{11}&a_{12}&\dots &a_{1n}\\a_{21}&a_{22}&\dots &a_{2n}\\\vdots &\vdots &\ddots &\vdots \\a_{n1}&a_{n2}&\dots &a_{nn}\end{array}}\right|} $  lub  $ {\displaystyle \det A=\det {\begin{bmatrix}a_{11}&a_{12}&\dots &a_{1n}\\a_{21}&a_{22}&\dots &a_{2n}\\\vdots &\vdots &\ddots &\vdots \\a_{n1}&a_{n2}&\dots &a_{nn}\end{bmatrix}}}$

#### Przykład:
${\displaystyle \det A={\begin{vmatrix}a_{11}&a_{12}\\a_{21}&a_{22}\end{vmatrix}}=a_{11}a_{22}-a_{12}a_{21}}$

${\displaystyle \det A={\begin{vmatrix}a_{11}&a_{12}&a_{13}\\a_{21}&a_{22}&a_{23}\\a_{31}&a_{32}&a_{33}\end{vmatrix}}=a_{11}a_{22}a_{33}+a_{21}a_{32}a_{13}+a_{31}a_{12}a_{23}-a_{21}a_{12}a_{33}-a_{11}a_{32}a_{23}-a_{31}a_{22}a_{13}}$

${\displaystyle \det A={\begin{vmatrix}2&4\\-1&3\end{vmatrix}}=2\cdot3-4\cdot(-1)} = 6 + 4 = 10 $





# Kroki obliczeń:

1. **Sprawdzenie zgodności wymiarów:**
W macierzy A liczba kolumn wynosi 2, a w macierzy B liczba wierszy wynosi 2 – warunek jest spełniony. Wynik mnożenia będzie macierzą o wymiarach (3,3) (czyli (liczba wierszy A, liczba kolumn B)).

2. **Obliczanie elementów macierzy wynikowej:**
Elementy iloczynu 𝐶 = 𝐴 ⋅ 𝐵 obliczamy:
### Przykładowo:
* C[0,0] = A[0,0] × B[0,0] + A[0,1] × B[1,0] = 2 × 4 + 3× (−1)= 8 − 3 = 5

* C[0,1] = A[0,0] × B[0,1] + A[0,1] × B[1,1] = 2 × 3 +3 × 0 = 6

* C[0,2] = A[0,0] × B[0,2] + A[0,1] × B[1,2] = 2 × 2 + 3 × 2 = 4 + 6 = 10

### Analogicznie obliczamy pozostałe elementy:

* C[1,0] = (−4) × 4 + 2 × (−1) = −16 − 2 = −18
* C[1,1] = (−4) × 3 + 2 × 0 = − 12
* C[1,2] = (−4) × 2 + 2 × 2 = −8 + 4= −4
* C[2,0] = 5 × 4 + 0 × (−1) = 20
* C[2,1] = 5 × 3 + 0 × 0 = 15
* C[2,2] = 5 × 2 + 0 × 2 = 10



### <a name='6'></a> Ćwiczenie 66
Podana jest poniższa tablica (macierz kwadratowa) $A$:

```
A = np.array([[-2, 0, 4],
              [5, 2, -1],
              [-4, 2, 4]])
```
Wyznacz wyznacznik macierzy $A$.

__Wskazówka:__ Użyj funkcji _np.linalg.det()_.


In [10]:
# tutaj wpisz rozwiązanie
A = np.array([[-2, 0, 4],
              [5, 2, -1],
              [-4, 2, 4]])

detA = np.linalg.det(A)
detA

51.999999999999986

Wyznacznik macierzy **𝐴** (dla macierzy 3x3) obliczamy według wzoru:

det(A)=a11(a22a33 − a23a32) − a12(a21a33 − a23a31 ) + a13(a21a32 − a22a31)

##### Podstawiając wartości:

* 𝑎11 = −2, 𝑎12 = 0, 𝑎13 = 4
* 𝑎21 = 5, a22 = 2, a23 = −1
* 𝑎31 = −4, a32 = 2, a33 = 4

##### Otrzymujemy:

* Pierwszy składnik: −2 × (2 × 4 − (−1) × 2) = −2 × (8 + 2) = −2 × 10 = −20
* Drugi składnik: 0 × (…) = 0
* Trzeci składnik: 4 × (5 × 2 − 2 × (−4)) = 4 × (10+8) = 4 × 18 = 72

Stąd:

    det(A) = −20 + 72 = 52

### <a name='7'></a> Ćwiczenie 67
Podana jest poniższa tablica (macierz kwadratowa) $A$:

```
A = np.array([[5, 8, 16],
              [4, 1, 8],
              [-4, 4, -11]])
```
Wyzancz wartości własne i odpowiadające im wektory własne macierzy $A$.

__Wskazówka:__ Użyj funkcji _np.linalg.eig()_.


In [15]:
# tutaj wpisz rozwiązanie
A = np.array([[5, 8, 16],
              [4, 1, 8],
              [-4, 4, -11]])
print(f"Macierz A: \n{A}\n")
eigenvalues, eigenvectors = np.linalg.eig(A)

print("Wartości własne:")
print(eigenvalues)

print("\nWektory własne (każda kolumna to wektor własny):")
print(eigenvectors)

Macierz A: 
[[  5   8  16]
 [  4   1   8]
 [ -4   4 -11]]

Wartości własne:
[-9.2462 -3.      7.2462]

Wektory własne (każda kolumna to wektor własny):
[[-0.6262 -0.8944  0.8902]
 [-0.3131  0.      0.4451]
 [ 0.7141  0.4472 -0.0976]]


1. **Wartości własne:**

* Uzyskałeś trzy wartości własne: −9.2462, −3 i 7.2462.
* Suma wartości własnych powinna być równa śladowi macierzy (czyli sumie elementów na głównej przekątnej). Dla macierzy

$A = \begin{bmatrix}5&8&16\\4&1&8\\-4&4&-11\end{bmatrix}$

ślad wynosi 5 + 1 + (−11) = −5. Sprawdźmy: −9.2462 + (−3) + 7.2462 ≈ −5, co jest zgodne z teorią.
* Iloczyn wartości własnych powinien być równy wyznacznikowi macierzy. W Twoim przypadku iloczyn (−9.2462) × (−3) × (7.2462) daje wartość zbliżoną do wyznacznika macierzy 𝐴.

### Przypomnienie
### Macierz odwrotna:

$A$ - macierz kwadratowa stopnia $n$. Macierz $A$ posiada macierz odwrotną, gdy istnieje macierz $B$, taka, że
$$AB = BA = I$$

$A = \begin{bmatrix}2&4\\-1&3\end{bmatrix}$

Funkcja ta zwraca:

* eigenvalues – jednowymiarową tablicę zawierającą wartości własne macierzy 𝐴.
* eigenvectors – tablicę, w której każda kolumna jest wektorem własnym odpowiadającym kolejno wyznaczonej wartości własnej.


### <a name='8'></a> Ćwiczenie 68
Podana jest poniższa tablica (macierz kwadratowa) $A$:

```
A = np.array([[5, 8, 16],
              [4, 1, 8],
              [-4, 4, -11]])
```
Wyznacz macierz odwrotną do macierzy $A$.

__Wskazówka:__ Użyj funkcji _np.linalg.inv()_.


In [16]:
# tutaj wpisz rozwiązanie
A = np.array([[5, 8, 16],
              [4, 1, 8],
              [-4, 4, -11]])
print(f"Macierz A: \n{A}\n")
invA = np.linalg.inv(A)
print("Macierz odwrotna do A:\n")
print(invA)

Macierz A: 
[[  5   8  16]
 [  4   1   8]
 [ -4   4 -11]]

Macierz odwrotna do A:

[[-0.2139  0.7562  0.2388]
 [ 0.0597  0.0448  0.1194]
 [ 0.0995 -0.2587 -0.1343]]


### Przypomnienie

### Ślad macierzy:

$${\displaystyle A={\begin{bmatrix}a_{11}&a_{12}&\dots &a_{1n}\\a_{21}&a_{22}&\dots &a_{2n}\\\vdots &\vdots &\ddots &\vdots \\a_{n1}&a_{n2}&\dots &a_{nn}\end{bmatrix}}}$$
$A$ - macierz wymiaru nxn

${\displaystyle \operatorname {tr} (A)=\sum _{i=1}^{n}a_{ii}=a_{11}+a_{22}+\dots +a_{nn}}$

#### Przykład:
$A = \begin{bmatrix}2&4\\-1&3\end{bmatrix}$

${\displaystyle \operatorname {tr} (A)=\sum _{i=1}^{2}a_{ii}=a_{11}+a_{22} = 2 + 3 = 5}$

### <a name='9'></a> Ćwiczenie 69
Podana jest poniższa tablica (macierz kwadratowa) $A$:

```
A = np.array([[5, 8, 16],
              [4, 1, 8],
              [-4, 4, -11]])
```
Wyznacz ślad macierzy kwadratowej $A$ (suma elementów na głównej przekątnej).

__Wskazówka:__ Użyj funkcji _np.trace()_.


In [17]:
# tutaj wpisz rozwiązanie
A = np.array([[5, 8, 16],
              [4, 1, 8],
              [-4, 4, -11]])

# trace(A) = 5 + 1 + (−11) = −5
traceA = np.trace(A)

print(f"Macierz A: \n{A}\n")
print(f"Ślad macierzy A: \n {traceA}\n")

Macierz A: 
[[  5   8  16]
 [  4   1   8]
 [ -4   4 -11]]

Ślad macierzy: 
 -5



#### Ślad macierzy to suma elementów znajdujących się na głównej przekątnej macierzy (czyli elementów, dla których indeks wiersza jest równy indeksowi kolumny).

### Przypomnienie
### Macierz transponowana:

Zamiana wierszy na kolumny i kolumny na wiersze.

#### Przykład 1:

$A = \begin{bmatrix}2&4\\-1&3\end{bmatrix}$,
$A^{T} = \begin{bmatrix}2&-1\\4&3\end{bmatrix}$

#### Przykład 2:
$A = \begin{bmatrix}2&1&-1\\0&-1&2\end{bmatrix}$,
$A^{T} = \begin{bmatrix}2&0\\1&-1\\-1&2\end{bmatrix}$


### Ćwiczenie 69 c.d.
Dokonaj transpozycji macierzy $A$.

In [18]:
# tutaj wpisz rozwiązanie
A_transposed = A.T

print("Macierz A:")
print(A)
print("\nTransponowana macierz A:")
print(A_transposed)

Macierz A:
[[  5   8  16]
 [  4   1   8]
 [ -4   4 -11]]

Transponowana macierz A:
[[  5   4  -4]
 [  8   1   4]
 [ 16   8 -11]]


### Wyjaśnienie:

* Transpozycja macierzy polega na zamianie wierszy z kolumnami.
* W powyższym przykładzie macierz A ma wymiary 3x3, więc jej transpozycja również będzie macierzą 3x3, gdzie element z pozycji A[i, j] przechodzi do pozycji A[j, i].

### <a name='10'></a> Ćwiczenie 70
Podane są dwie dwuwymiarowe tablice numpy $A$ oraz $B$:   
$A$ - macierz wymiaru $(4, 2)$  
$B$ - macierz wymiaru $(1, 8)$

```
A = np.array([[2, 0],
              [4, 2],
              [5, 3],
              [4, 2]])
B = np.array([[4, 0, 2, 1, 1, 0, 2, 9]])
```
Przekształć tablicę $B$ tak, aby można było dokonać mnożenia macierzowego $A \cdot B$. Wykonaj to mnożenie.

__Oczekiwany wynik:__


```
array([[ 8,  0,  4,  2],
       [18,  0, 12, 22],
       [23,  0, 16, 32],
       [18,  0, 12, 22]])
```




In [23]:
# tutaj wpisz rozwiązanie
A = np.array([[2, 0],
              [4, 2],
              [5, 3],
              [4, 2]])
B = np.array([[4, 0, 2, 1, 1, 0, 2, 9]])

print(f"Macierz A: \n {A} \n")

# Przekształcenie macierzy B:
# Używamy funkcji reshape(), aby zmienić kształt B z (1,8) na (2,4):
B_reshaped = B.reshape(2, 4)
print(f"Przekształcona macierz B (kształt (2,4)) \n{B_reshaped}\n")

# Mnożenie macierzy
# Teraz, gdy A ma kształt (4,2) a B_new (2,4), mnożenie A ⋅ B_new jest możliwe i da macierz wynikową o wymiarach (4,4).
# Mnożenie macierzy wykonujemy za pomocą funkcji np.dot() lub operatora @
C = np.dot(A, B_reshaped)
print(f"Mnozenie macierzy A i przekształconej macierzy B: \n{C}\n")

# Obliczenie elementów macierzy wynikowej:
# iloczyn macierzy (4,2) i (2,4) daje macierz (4,4)


Macierz A: 
 [[2 0]
 [4 2]
 [5 3]
 [4 2]] 

Przekształcona macierz B (kształt (2,4)) 
[[4 0 2 1]
 [1 0 2 9]]

Mnozenie macierzy A i przekształconej macierzy B: 
[[ 8  0  4  2]
 [18  0 12 22]
 [23  0 16 32]
 [18  0 12 22]]

